[DependencyInjection] Autoconfigurable attributes on methods, properties and parameters

ruudk opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      |no
| New feature?  | yes
| Deprecations? |no 
| Tickets       | 
| License       | MIT
| Doc PR        | 

## Introduction

https://github.com/symfony/symfony/pull/39897 introduced the possibility auto configure classes that were annotated with attributes:
```php
$container->registerAttributeForAutoconfiguration(
    MyAttribute::class,
    static function (ChildDefinition $definition, MyAttribute $attribute, \ReflectionClass $reflector): void {
        $definition->addTag('my_tag', ['some_property' => $attribute->someProperty]);
    }
);
```

This works great. But it only works when the attribute is added on the class. 

With this PR, it's now possible to also auto configure methods, properties and parameters.

## How does it work?

Let's say you have an attribute that targets classes and methods like this:

```php
#[Attribute(Attribute::TARGET_CLASS | Attribute::TARGET_METHOD | Attribute::TARGET_PROPERTY)]
final class MyAttribute
{
}
```

You have two services that use them:
```php
#[MyAttribute]
class MyService {
}

class MyOtherService {
    #[MyAttribute]
    public function myMethod() {}
}
```

You can now use `registerAttributeForAutoconfiguration` in your extension, together with a union of the types that you want to seach for. In this example, the extension only cares for classes and methods, so it uses  `\ReflectionClass|\ReflectionMethod $reflector`:
```php
final class MyBundleExtension extends Extension
{
    public function load(array $configs, ContainerBuilder $container) : void
    {
        $container->registerAttributeForAutoconfiguration(
            MyAttribute::class,
            static function (ChildDefinition $definition, MyAttribute $attribute, \ReflectionClass|\ReflectionMethod $reflector) : void {
                $args = [];
                if ($reflector instanceof \ReflectionMethod) {
                    $args['method'] = $reflector->getName();
                }
                $definition->addTag('my.tag', $args);
            }
        );
    }
}
```

This will tag `MyService` with `my.tag` and it will tag `MyOtherService` with `my.tag, method: myMethod`

If the extension also wants to target the properties that are annotated with attributes, it can either change the union to `\ReflectionClass|\ReflectionMethod|\ReflectionProperty $reflector` or it can just use `\Reflector $reflector` and do the switching in the callable.

## Another example

Let's say you have an attribute like this:
```php
#[Attribute(Attribute::TARGET_CLASS)]
final class MyAttribute
{
}
```

and you use it like this:
```php
$container->registerAttributeForAutoconfiguration(
    MyAttribute::class,
    static function (ChildDefinition $definition, MyAttribute $attribute, \ReflectionClass|\ReflectionMethod $reflector) : void {
        $definition->addTag('my.tag');
    }
);
```

you'll get an error saying that `ReflectionMethod` is not possible as the attribute only targets classes.

ruudk committed:
[DependencyInjection] Autoconfigurable attributes on methods, properties and parameters

stof commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1309,7 +1309,7 @@ public function registerForAutoconfiguration(string $interface)
      * Registers an attribute that will be used for autoconfiguring annotated classes.
      *
-     * The configurator will receive a ChildDefinition instance, an instance of the attribute and the corresponding \ReflectionClass, in that order.
+     * The configurator will receive a ChildDefinition instance, an instance of the attribute and the corresponding \ReflectionClass or \ReflectionMethod, in that order.
```
This is a BC break, as a callback typehinting `ReflectionClass` will break when upgrading to 5.4. We cannot widen the parameter we pass to a callable.

ruudk replied:
We could add `ReflectionMethod` as 4rd nullable parameter?

ruudk replied:
I pushed a change where a 4rd `ReflectionMethod` parameter is added.

stof replied:
that could still be a BC break: the callable that is not aware of ReflectionMethod would be called for method-level attributes without handling them properly (as the attribute is not on the ReflectionClass)

nicolas-grekas replied:
https://github.com/symfony/symfony/pull/39897#discussion_r561148511 is where I suggested passing the reflection object to the configurator. And as you'll see there, I was already anticipating that we could pass any Reflector there, not only a ReflectionClass.

I think we can and should pass the ReflectionMethod directly. In terms of BC, we should first update the comment in 5.3 to make the intent clear, and then, we could skip caring: any TypeErrror that might result from a too restrictive type-hint will be caught at compile-time. Aka it's trivial to spot. If we want a smoother path, we could check via reflection that passing a ReflectionMethod would be fine and skip if not. But I'm not sure this is worth the extra code and complexity.

ruudk replied:
~@nicolas-grekas Shall I partially revert https://github.com/symfony/symfony/pull/42039/commits/6a917af5b92c59d6710806ff6413106a5ab7f390 ?~ nvm

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -20,6 +20,8 @@
 use Psr\Http\Client\ClientInterface;
 use Psr\Log\LoggerAwareInterface;
+use ReflectionMethod;
+use Reflector;
```
We don't use use statements for classes of the global namespace

stof commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -48,6 +48,13 @@ protected function processValue($value, bool $isRoot = false)
+        foreach ($reflector->getMethods() as $method) {
+            foreach ($method->getAttributes() as $attribute) {
+                if ($configurator = $autoconfiguredAttributes[$attribute->getName()] ?? null) {
+                    $configurator($conditionals, $attribute->newInstance(), $method);
```
Using all existing callable with methods might have unintended effects if they were expecting to run on classes only. Maybe we should have a separate registration for method-level autoconfiguration attributes

nicolas-grekas replied:
That's not really "any callable", but "any callable bound to the attribute", which itself has to opt-in for being accepted on methods. That might be good enough and we might not need more.

nicolas-grekas commented:
Scanning all methods of all autoconfigured services might be a perf hog on real apps.
We should understand the impact this could have before going this way.
Can you run some benchmark and report your findings please?
IIRC, Doctrine does it with an extra annotation on the class, basically saying "have a look at methods".

ruudk commented:
@nicolas-grekas Not sure how to _properly_ benchmark this, but I used a `Stopwatch` in my _large application_ and these are the results:
Class attributes took 3.000000 ms for 4469 classes
Method attributes took 18.500000 ms for 4469 classes and total of 41920 methods

<details>
  <summary>Benchmark code</summary>
  
  ```php
    private Stopwatch $stopwatch;
    private int $classes = 0;
    private int $methods = 0;

    public function __destruct()
    {
        dump(sprintf(
            'Class attributes took %f ms for %d classes',
            $this->stopwatch->getEvent('class')->getDuration(),
            $this->classes
        ));
        dump(sprintf(
            'Method attributes took %f ms for %d classes and total of %d methods',
            $this->stopwatch->getEvent('method')->getDuration(),
            $this->classes,
            $this->methods
        ));
    }

    protected function processValue($value, bool $isRoot = false)
    {
        // ...
        $this->stopwatch->start('class');
        ++$this->classes;
        foreach ($reflector->getAttributes() as $attribute) {
            if ($configurator = $autoconfiguredAttributes[$attribute->getName()] ?? null) {
                $configurator($conditionals, $attribute->newInstance(), $reflector);
            }
        }
        $this->stopwatch->stop('class');
        $this->stopwatch->start('method');
        foreach ($reflector->getMethods() as $method) {
            ++$this->methods;
            foreach ($method->getAttributes() as $attribute) {
                if ($configurator = $autoconfiguredAttributes[$attribute->getName()] ?? null) {
                    $configurator($conditionals, $attribute->newInstance(), $method);
                }
            }
        }
        $this->stopwatch->stop('method');
        // ...
    }
  ```
</details>

Not sure if this is acceptable and how it can be improved.

nicolas-grekas commented:
Thanks for the bench, this looks acceptable!

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1309,7 +1309,7 @@ public function registerForAutoconfiguration(string $interface)
      * Registers an attribute that will be used for autoconfiguring annotated classes.
      *
-     * The configurator will receive a ChildDefinition instance, an instance of the attribute and the corresponding \ReflectionClass, in that order.
+     * The configurator will receive a ChildDefinition instance, an instance of the attribute, the corresponding \ReflectionClass, and optional \ReflectionMethod, in that order.
```
```
     * The configurator will receive a ChildDefinition instance, the attribute instance, and the corresponding
     * \ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter, in that order.
```

ruudk replied:
Changed, but `ReflectionParameter` is still not implemented.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -48,6 +48,13 @@ protected function processValue($value, bool $isRoot = false)
                 $configurator($conditionals, $attribute->newInstance(), $reflector);
             }
         }
+        foreach ($reflector->getMethods() as $method) {
```
Let's scan only ` ReflectionMethod::IS_PUBLIC`, and let's skip static methods (same for properties.)

ruudk replied:
Done

nicolas-grekas reviewed (commented):
I'd suggest extending this to properties and parameters.

In order to play safe and optimize perf a bit, I'd suggest doing some reflection in the `process()` method to classify configurators in 4 groups: the ones that work on classes (ie the ones that have only two arguments or that have a 3rd arg that accept ReflectionClass), the ones that work on methods, properties, and parameters. If a group is empty, we would skip scanning for the corresponding type of attributes.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -551,8 +551,12 @@ public function load(array $configs, ContainerBuilder $container)
+        $container->registerAttributeForAutoconfiguration(AsEventListener::class, static function (ChildDefinition $definition, AsEventListener $attribute, \ReflectionClass $reflectionClass, ?\ReflectionMethod $reflectionMethod): void {
+            $tagAttributes = get_object_vars($attribute);
+            if ($reflectionMethod) {
+                $tagAttributes['method'] = $reflectionMethod->getName();
```
we should throw if the attribute is on a method and also defines a "method" entry

ruudk replied:
Done

ruudk commented:
> In order to play safe and optimize perf a bit, I'd suggest doing some reflection in the process() method to classify configurators in 4 groups: the ones that work on classes (ie the ones that have only two arguments or that have a 3rd arg that accept ReflectionClass), the ones that work on methods, properties, and parameters. If a group is empty, we would skip scanning for the corresponding type of attributes.

@nicolas-grekas I applied this feedback and pushed a WIP commit.

Since Attributes only work on PHP 8 I thought it would be nice to support Union types when you want to register an attribute for multiple methods. 

So for my example of the `AsEventListener` I want to typehint that with `ReflectionClass | ReflectionMethod`. 

This works, but it will produce a fatal error on PHP 7.4 because it doesn't support unions there. How can we fix that? 
Or should we target the `AsEventListener` change for Symfony 6 as that requires PHP 8 and up?

I'm also not sure how `ReflectionParameter` should be implemented. Should that work on all methods? Or only promoted parameters in the constructor? Aren't those already picked by `ReflectionProperty`?

## Todo

- [x] How to handle `AsEventListener` on method change? Move to Symfony 6 or find alternative for Union?
- [x] Add and fix all tests
- [x] `ReflectionParameter`

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -551,8 +551,16 @@ public function load(array $configs, ContainerBuilder $container)
 
-        $container->registerAttributeForAutoconfiguration(AsEventListener::class, static function (ChildDefinition $definition, AsEventListener $attribute): void {
-            $definition->addTag('kernel.event_listener', get_object_vars($attribute));
+        // @todo this will produce a fatal error on PHP 7.4 because it doesn't understand unions
```
we should remove the type then

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -551,8 +551,16 @@ public function load(array $configs, ContainerBuilder $container)
+            $tagAttributes = get_object_vars($attribute);
+            if ($reflector instanceof \ReflectionMethod) {
+                if (isset($tagAttributes['method'])) {
+                    throw new LogicException('Cannot set "method" when using AsEventListener on method.');
```
we need to tell the affected class+method to help ppl fix the mistake

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -11,6 +11,7 @@
 
 namespace Symfony\Component\DependencyInjection\Compiler;
 
+use LogicException;
```
should be removed and `\LogicException` should be used inline, per our code style

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -20,12 +21,50 @@
+                }
+            } else {
+                if ($parameterType === null) {
+                    throw new LogicException('Callable parameter 3 should have a type.');
```
too restrictive on PHP7
we should instead consider that all types are accepted

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -20,12 +21,50 @@
+                if ($parameterType === null) {
+                    throw new LogicException('Callable parameter 3 should have a type.');
+                }
+                $types[] = $parameterType->getName();
```
to get rid of the psalm error, we should use
```php
if ($parameterType instanceof \ReflectionUnionType) {
//...
} elseif ($parameterType instanceof \ReflectionNamedType) {
//...
} else {
// consider that all types are accepted
}

nicolas-grekas replied:
thinking again about that: I'm fine with the exception!
```
} else {
   throw
}
```

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -20,12 +21,50 @@
 {
+    private $classAttributes = [];
+    private $methodAttributes = [];
+    private $propertyAttributes = [];
```
I'd suggest naming those: *Configurators

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -35,21 +74,46 @@ protected function processValue($value, bool $isRoot = false)
-                $configurator($conditionals, $attribute->newInstance(), $reflector);
+        $conditionals = $instanceof[$classReflector->getName()] ?? new ChildDefinition('');
+
+        if (count($this->classAttributes) > 0) {
```
`if ($this->classConfigurators) {` (same for the others)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -35,21 +74,46 @@ protected function processValue($value, bool $isRoot = false)
+        }
+
+        if (count($this->methodAttributes) > 0) {
+            foreach ($classReflector->getMethods(\ReflectionMethod::IS_PUBLIC & ~\ReflectionMethod::IS_STATIC) as $methodReflector) {
```
I think the non-static filter doesn't work and should be replaced by an explicit `->isStatic()` call
we should also skip when `isConstructor()` and when `isDestructor()`.

in this loop, we should analyze parameters when needed

we should also analyze the parameters of the constructor, but to get the corresponding reflector, we should use `$this->getConstructor($value, false)`

ruudk replied:
Done

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -35,21 +74,46 @@ protected function processValue($value, bool $isRoot = false)
+            }
+        }
+
+        if (count($this->methodAttributes) > 0) {
```
if ($this->methodConfigurators || $this->parameterConfigurators) {

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -35,21 +74,46 @@ protected function processValue($value, bool $isRoot = false)
-            $instanceof[$reflector->getName()] = $conditionals;
+
+        if (count($this->propertyAttributes) > 0) {
+            foreach ($classReflector->getProperties(~\ReflectionProperty::IS_STATIC) as $propertyReflector) {
```
`ReflectionProperty::IS_PUBLIC`, and the check for non-static props should be done by calling the method

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -35,21 +74,46 @@ protected function processValue($value, bool $isRoot = false)
+            }
+        }
+
+        // @todo parameters... on what? Constructor or all methods?
```
constructor and all methods, see previous comment

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -551,8 +551,16 @@ public function load(array $configs, ContainerBuilder $container)
-        $container->registerAttributeForAutoconfiguration(AsEventListener::class, static function (ChildDefinition $definition, AsEventListener $attribute): void {
-            $definition->addTag('kernel.event_listener', get_object_vars($attribute));
+        // @todo this will produce a fatal error on PHP 7.4 because it doesn't understand unions
+        $container->registerAttributeForAutoconfiguration(AsEventListener::class, static function (ChildDefinition $definition, AsEventListener $attribute, \ReflectionClass|\ReflectionMethod $reflector): void {
```
let's use eval() to define the function:

```php
if (\PHP_VERSION_ID >= 80000) {
    $container->registerAttributeForAutoconfiguration(AsEventListener::class, eval('return static function...'
```

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -20,12 +20,59 @@
+            } elseif ($parameterType instanceof \ReflectionNamedType) {
+                $types[] = $parameterType->getName();
+            } else {
+                $types = [
```
let's throw instead, I changed my mind :)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -35,21 +82,81 @@ protected function processValue($value, bool $isRoot = false)
+        if ($this->methodAttributeConfigurators || $this->parameterAttributeConfigurators) {
+            $constructorReflector = $this->getConstructor($value, false);
+            if ($constructorReflector) {
+                foreach ($constructorReflector->getAttributes() as $attribute) {
```
We should remove this loop: attributes on constructors don't make sense here, ppl should use class-level attributes instead

ruudk replied:
What about promoted properties?

nicolas-grekas replied:
they will appear as both properties and params

ruudk replied:
You're right!

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -35,21 +82,81 @@ protected function processValue($value, bool $isRoot = false)
+                    continue;
+                }
+
+                foreach ($methodReflector->getAttributes() as $attribute) {
```
`if ($this->methodAttributeConfigurators)`

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1309,7 +1309,12 @@ public function registerForAutoconfiguration(string $interface)
+     * The configurator will receive (in this order):
+     *  - a ChildDefinition instance
+     *  - the attribute instance
+     *  - the corresponding reflector (\ReflectionClass, \ReflectionMethod, \ReflectionProperty, \ReflectionParameter)
```
let's use `|` instead of `, ` in the enumeration, that's more clear IMHO

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1309,7 +1309,12 @@ public function registerForAutoconfiguration(string $interface)
+     *  - the attribute instance
+     *  - the corresponding reflector (\ReflectionClass, \ReflectionMethod, \ReflectionProperty, \ReflectionParameter)
+     *
+     * If your attribute can be used on multiple targets, you can use a union type to support it.
```
`If an attribute can be used on multiple targets, a union type should be used.`

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Tests/Compiler/IntegrationTest.php:
```diff
@@ -12,6 +12,10 @@
+use ReflectionClass;
+use ReflectionMethod;
+use ReflectionParameter;
+use ReflectionProperty;
```
no `use` for the root namespace

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Tests/Compiler/IntegrationTest.php:
```diff
@@ -699,6 +708,88 @@ static function (Definition $definition, CustomAutoconfiguration $attribute) {
+        $container->registerAttributeForAutoconfiguration(
+            CustomAnyAttribute::class,
+            eval(<<<'PHP'
+            return static function (\Symfony\Component\DependencyInjection\ChildDefinition $definition, \Symfony\Component\DependencyInjection\Tests\Fixtures\Attribute\CustomAnyAttribute $attribute, \ReflectionClass | \ReflectionMethod | \ReflectionProperty | \ReflectionParameter $reflector) {
```
no spaces around `|`

ruudk replied:
Weird, because fabbot requires me to do this (in another PR)? 

https://fabbot.io/patch/symfony/symfony/42051/f38cc2e8c03ec6215d28c97e7ec534b2d9be6010/cs.diff

nicolas-grekas replied:
yes,fabbot has an issue right now, we're waiting for a release new of php-cs-fixer

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -20,12 +20,54 @@
+            } elseif ($parameterType instanceof \ReflectionNamedType) {
+                $types[] = $parameterType->getName();
+            } else {
+                throw new \LogicException('Callable passed to registerAttributeForAutoconfiguration() should have a type for parameter #3. Use one or more of the following types: \ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter.');
```
let's use LogicException from the component's namespace
let's also tell file + line info to help spot the source of the issue

ruudk replied:
Used the correct `LogicException` now. But I'm not sure how to get the file and line info as this is executed in the `CompilerPass` and not where it's defined (the call to `registerAttributeForAutoconfiguration`)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -35,21 +77,77 @@ protected function processValue($value, bool $isRoot = false)
+
+        if ($this->methodAttributeConfigurators || $this->parameterAttributeConfigurators) {
+            $constructorReflector = $this->getConstructor($value, false);
+            if ($constructorReflector) {
```
`if ($this->parameterAttributeConfigurators && $constructorReflector = $this->getConstructor($value, false)) {`

ruudk replied:
Done, made it a bit smarter.

nicolas-grekas reviewed (approved):
LGTM, I only have 2 minor comments.
Thanks for completing this quickly!

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -14,18 +14,61 @@
+            } elseif ($parameterType instanceof \ReflectionNamedType) {
+                $types[] = $parameterType->getName();
+            } else {
+                throw new LogicException('Callable passed to registerAttributeForAutoconfiguration() should have a type for parameter #3. Use one or more of the following types: \ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter.');
```
$callableReflector->getFileName()+getStartLine()?

ruudk replied:
Works! 😍

Tobion commented:
I'd like to see some use-cases for each way of auto-configuring attributes. This would also be required for the docs at the end anyway.

ruudk commented:
The reason for creating this PR was to be able to have attributes on methods. In our application we make use of event subscribers that act on events published on our event bus ([SimpleBus](https://github.com/SimpleBus/SimpleBus)). Some subscribers need to be run asynchronous. Some subscribers subscribe to a lot of different events. 

We introduced a `Asynchronous` attribute that lets us specify queues and delays. 

```php
final class ProcessManager
{
    #[Asynchronous(delayInMinutes: 5)]
    public function cleanup(CartOpenedEvent $event) {
      // Do something 5 minutes after cart was opened
    }
    
    #[Asynchronous(queue: 'cart')]
    public function onCartClosed(CartClosedEvent $event) {
      // Cart was closed, remove it.
    }
}
```

After receiving feedback I also introduced support for properties and parameters. But I don't really need that.

Tobion commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1309,7 +1309,12 @@ public function registerForAutoconfiguration(string $interface)
      * Registers an attribute that will be used for autoconfiguring annotated classes.
      *
-     * The configurator will receive a ChildDefinition instance, an instance of the attribute and the corresponding \ReflectionClass, in that order.
+     * The configurator will receive (in this order):
```
I'd suggest to add a psalm/phpstan signature for the configurator callable

ruudk replied:
Applied your feedback.

Tobion commented:
I think the feature and implementation looks pretty clean and flexible.
I just fear that we add more and more of autoconfiguration options that duplicate alot of functionality.
I'm trying to list the possibilities of autoconfiguration:
- based on marker interfaces purely for autoconfiguration like https://github.com/symfony/messenger/blob/5.3/Handler/MessageHandlerInterface.php and https://github.com/symfony/symfony/blob/5.3/src/Symfony/Bundle/FrameworkBundle/Routing/RouteLoaderInterface.php
- autoconfiguration per directory like https://symfony.com/blog/new-in-symfony-4-4-simpler-event-listeners
- autoconfiguration based on parameter types (e.g. message handler, event listener)
- autoconfiguration based on subscribers with the logic in php code like EventSubscriberInterface and ServiceSubscriberInterface
- autoconfiguration based on attributes in classes, properties, methods and parameters

ruudk commented:
@Tobion I think 1 and 4 are the same thing. 1 doesn't hold any information, just a tag. And 4 require the implementer to return some data. But same principle?

The list that you created is what I love about Symfony. I can decide to use instanceof/interfaces/markers, or just configure based on directory, or introduce new attributes where they make sense.

ruudk commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1309,7 +1309,12 @@ public function registerForAutoconfiguration(string $interface)
+     *  - the attribute instance
+     *  - the corresponding reflector (\ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter)
+     *
+     * If an attribute can be used on multiple targets, a union type should be used.
```
```suggestion
     * If an attribute can be used on multiple targets, a union type should be used.
     *
     * @param callable(ChildDefinition, object $attribute, \ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter):void $configurator
```

@Tobion Like this?

Tobion replied:
Yes. Then there is no need for the proposa "The configurator will receive (in this order): ...".
Would be even better to tell static analysis that the `$attribute` param is an instance of `$attributeClass`. Should also be possible.

Tobion replied:
also `@param class-string $attributeClass` would be good

ruudk replied:
Applied your feedback.

Tobion commented:
I wouldn't say marker interfaces and subscribers are the same thing. marker interfaces have only been introduces for autoconfiguration in DI and no other purpose. EventSubscriberInterface on the other hand is actually used in the EventDispatcher component. They have the same purpose yes. But so do the new autoconfigurable attributes. So there are now at least 4 ways to register an event lister:
- using explicit EventSubscriberInterface
- using autoconfiguration of event listeners by param types
- using explicit wiring in DI
- using AsEventListener attribute

Tobion commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1309,7 +1309,10 @@ public function registerForAutoconfiguration(string $interface)
+     * If an attribute can be used on multiple targets, a union type should be used for
+     * the \Reflector parameter. Oterwise specify one \Reflector type.
+     *
+     * @param callable(ChildDefinition, class-string $attribute, \ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter):void $configurator
```
`class-string $attribute` is not correct. $attributeClass is a class-string

nicolas-grekas replied:
this should be `object $attribute`

ruudk replied:
Fixed. Added `@param class-string $attributeClass` as well.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1309,7 +1309,11 @@ public function registerForAutoconfiguration(string $interface)
      *
-     * The configurator will receive a ChildDefinition instance, an instance of the attribute and the corresponding \ReflectionClass, in that order.
+     * If an attribute can be used on multiple targets, a union type should be used for
+     * the \Reflector parameter. Oterwise specify one \Reflector type.
```
Otherwise

ruudk replied:
Reverted the changes of today.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1309,7 +1309,11 @@ public function registerForAutoconfiguration(string $interface)
+     * the \Reflector parameter. Oterwise specify one \Reflector type.
+     *
+     * @param class-string $attributeClass
+     * @param callable(ChildDefinition, object $attribute, \ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter):void $configurator
```
There is an issue with this annotation: if one type hints `ReflectionClass $foo`, psalm will complain that the type hint too narrow. I'm not sure we should use this actually.

ruudk replied:
Reverted the changes of today.

Tobion replied:
Please don't revert this. Nicolas has a point. But we can just use `\Reflector` as typehint then and clarify in the doc that this is either `\ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter` or a union of them depending on the attribute target.

ruudk replied:
@tobion Applied your feebback.

Tobion commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -551,9 +551,23 @@ public function load(array $configs, ContainerBuilder $container)
+        if (\PHP_VERSION_ID >= 80000) {
+            $container->registerAttributeForAutoconfiguration(AsEventListener::class, eval(
+                <<<'PHP'
+                return static function (\Symfony\Component\DependencyInjection\ChildDefinition $definition, \Symfony\Component\EventDispatcher\Attribute\AsEventListener $attribute, \ReflectionClass|\ReflectionMethod $reflector): void {
```
wrapping this in `eval` is pretty ugly. this is only due to the union type in `\ReflectionClass|\ReflectionMethod $reflector` right? why not just typhint `\Reflector $reflector` here? using a union type and sacrificing php 7 parsing for this does not seem worth it.

ruudk replied:
The union type hint is done so that we can optimize. If we would add this on Reflector we'd have to loop through everything as we don't know what the attribute should work on.

Unless we use reflection on the attribute to find the target? And then use that information.

Tobion replied:
Using the attribute target seems like to correct solution to me. Relying on the argument type is not reliable.

nicolas-grekas replied:
Please explain why you think so? Because to me this is reliable.

nicolas-grekas replied:
To be more explicit, I think using the attribute target is less flexible. We'd better make configurators explicit about what the accept. The proposed way to express this is via this 3rd argument, and I see only advantages in doing so, no drawbacks.

The `eval()` here is a non issue.

Tobion replied:
Because they would need to make sure the type is in sync with the attribute target themselves. If they use the wrong type, the configurator might just be ignored without even an error. Pretty bad DX to me.

nicolas-grekas replied:
Same when using the attribute on eg private, protected methods when we read them only on public ones.
We cannot prevent every mistake, especially not when it'd mean doing extensive CPU processing.

Tobion replied:
When would it be useful to have an attribute with class target and someone accidentlly typehints ReflectionMethod? There is no error or anything. Debugging hell.

nicolas-grekas replied:
`eval()` now removed

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1309,7 +1309,12 @@ public function registerForAutoconfiguration(string $interface)
+     *
+     * If an attribute can be used on multiple targets, a union type can be used based on above types.
+     *
+     * @param callable(ChildDefinition $definition, object $attribute, \Reflector $reflector): void $configurator
```
Psalm will also complain with this, when a userland configurator will use eg `ReflectionMethod $r`...

Tobion replied:
then they can ignore the error or they \assert($reflector instanceof ReflectionMethod) inside the callable instead. To me autocomplete and proper signature documentation is the right thing to do.

Also this doc should be
```
@template T
@param class-string<T> $attributeClass
@param callable(ChildDefinition $definition, T $attribute, \Reflector $reflector): void $configurator
```

nicolas-grekas replied:
As @ruudk mentioned earlier, the type-hint is used as a filter to skip scanning attributes on eg parameters when no attribute is expected on them. It's true that we could improve the filter and also consider the attributes on the attribute classes, but I'm not sure we need the extra code, because as I wrote in https://github.com/symfony/symfony/pull/42039#discussion_r670431074, relying on attribute's flags is less flexible than relying on the argument's type hint: it creates an unnecessary coupling between the configurators and the attributes their deal with.

Anyway, I'm fine with the proposed phpdoc, provided we throw an explanatory exception when `Reflector` is kept as a type-hint.

ruudk replied:
Added the proposed phpdoc and added a guard to prevent usage of `Reflector` argument.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -14,18 +14,65 @@
+                throw new LogicException(sprintf('Parameter "$%s" in callable passed to registerAttributeForAutoconfiguration() in "%s" on line "%d" should have a type. Use one or more of the following types: \ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter.', $reflectorParameter->getName(), $callableReflector->getFileName(), $callableReflector->getStartLine()));
+            }
+
+            if (\in_array(\Reflector::class, $types, true)) {
```
thanks for adding this

I'm actively thinking about @Tobion's feedback and I think we can accommodate both our approaches by filtering based on both the attributes' targets and the allowed type hints.

Note that I don't know how to read the `#[Attribute]` attribute on attribute classes...

Can you figure out how? Anyone else?

nicolas-grekas replied:
This should do it:
```php
try {
    $r = new \ReflectionClass($attributeName);
} catch (\ReflectionException $e) {
    continue;
}

$targets = $r->getAttributes(\Attribute::class)[0] ?? 0;
$targets = $targets ? $targets->getArguments()[0] ?? -1 : 0;
```

nicolas-grekas replied:
we could even throw when $targets and the type hints of an argument have an empty intersection

ruudk replied:
I think it's too complicate to support both.

My suggestion would be:
* Read targets from Attribute with the above snippet, and use that
* Typehint in phpdoc is `\Reflector` but can also be made specific (e.g. `\ReflectionMethod`) or a union type in user code. It's not read by Symfony. 

@nicolas-grekas @Tobion If you agree with this I will update the PR.

nicolas-grekas replied:
I'm AFK for two weeks and I'd like to see how supporting both would work. I can give it a try when I'll be back!

ruudk replied:
@nicolas-grekas If you can give it a try, please do. Thanks 🙏

ruudk replied:
@nicolas-grekas Hope you had a good time hiking in Greece. Curious if you already have an idea how to solve this?

nicolas-grekas reviewed (commented):
@ruudk I force-pushed on your fork with the filter we talked about.
An exception will now be thrown if eg a `ReflectionClass` type is used with an attribute that cannot target a type, and one can use the generic `Reflector` type to get reflectors for all positions supported by an attribute.

@Tobion this should solve your concerns.

ruudk commented:
Much better!

fabpot commented:
@ruudk Can you work on a PR for the docs, or at least describe the feature with examples in the PR description? Thank you.

Tobion commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -14,18 +14,66 @@
             return;
         }
 
+        foreach ($container->getAutoconfiguredAttributes() as $attributeName => $callable) {
```
i think this block of code could be extracted into a separate private method to make the code more clear

ruudk replied:
Will it really? I don't agree, but if you give me a name for the method I'll do the change.

Tobion commented on src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php:
```diff
@@ -35,21 +83,74 @@ protected function processValue($value, bool $isRoot = false)
-        foreach ($reflector->getAttributes() as $attribute) {
-            if ($configurator = $autoconfiguredAttributes[$attribute->getName()] ?? null) {
-                $configurator($conditionals, $attribute->newInstance(), $reflector);
+        $conditionals = $instanceof[$classReflector->getName()] ?? new ChildDefinition('');
```
I would also put all of this new code into a new method

ruudk replied:
Will it really make it more clear? I don't agree, but if you give me a name for the method I'll do the change.

Tobion reviewed (approved):
I like it now. It will probably have some performance impact on container building. But I think using attributes for this might be the future. So there is no way around it. This will probably replace the marker interface solutions currently in place when there are such attributes as alternatives.

ruudk commented:
@fabpot I updated the PR description and added a few examples. Hope this helps!

fabpot commented:
> @fabpot I updated the PR description and added a few examples. Hope this helps!

That's perfect, thank you.

fabpot commented:
Thank you @ruudk.

ruudk commented:
@nicolas-grekas This PHPStan annotation doesn't work as we expect it to work
https://github.com/symfony/symfony/blob/446686ab06e8535739c51639ebd9fcc3b60f4b70/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L1337

See https://github.com/phpstan/phpstan/issues/6200

How can we solve this?

```php
$container->registerAttributeForAutoconfiguration(
            WithFlow::class,
            static function (ChildDefinition $definition, WithFlow $attribute, ReflectionMethod $reflector) : void {
                $definition->addTag(ServiceTag::WITH_FLOW, [
                    'flow' => $attribute->getFlowName(),
                ]);
            }
        );
```

gives this error:
```
phpstan: Parameter #2 $configurator of method 
Symfony\Component\DependencyInjection\ContainerBuilder::registerAttributeForAutoconfiguration() 
expects callable(Symfony\Component\DependencyInjection\ChildDefinition, WithoutFlow, Reflector): 
void, Closure(Symfony\Component\DependencyInjection\ChildDefinition, WithoutFlow, 
ReflectionMethod): void given.
```

nicolas-grekas commented:
`Closure(Reflector)|Closure(ReflectionMethod)|Closure(ReflectionMethod|ReflectionProperty)|etc`

I'm not sure phpstan supports any simple way to express the closures we support here. Ignoring the error might be just fine.
